docker-compose.yml

```diff
@@ -13,6 +13,19 @@ services:
       - config/docker.env
     stdin_open: true
     tty: true
+    depends_on:
+      - database
+
+  assets:
+    build: .
+    command: bin/webpack-dev-server
+    volumes:
+      - .:/usr/src/app
+      - /usr/src/app/node_modules
+    env_file:
+      - config/docker.env
+    stdin_open: true
+    tty: true
```

Comment:
Doesn't this just end up running the `webpack-dev-server` in a separate container that is completely isolated from the web application (and therefore not doing anything)?

Reply:
I did try to set up the webpack dev server as a separate service by opening port 3035 and setting the `WEBPACKER_DEV_SERVER_HOST` variable on both the webpack and web services, but I couldn't get it to work. If you've had success with that, please give it a shot in getting it set up. But I think just running the webpack server in the same container as the web application is not a big deal.

Reply:
It shares the same mounted volume...

Reply:
Ah, but that's not the point because this is actually serving assets itself. Sorry, my misunderstanding.

Reply:
It's actually using an anonymous volume...

Reply:
> But I think just running the webpack server in the same container as the web application is not a big deal.

It wouldn't be if it were started automatically. When one has to stop the web container to run DB migrations, it's actually quite a drag to have to start everything up again in multiple terminal panes, which is what I was trying to address.

Reply:
If you use `exec`, you don't need to stop the terminal to run the DB migrations. You just can run one container the whole time, which is also helpful for not having to immediately rebuild your image if someone adds a new gem. 

What benefit do you see in using `run` vs `exec`?

Reply:
If you'd like to keep webpack-dev-server in a separate `assets` container service but still start it automatically when the `web` container is run, you could add `assets` to the `web` service's `depends_on:` key.

```
depends_on:
  - database
  - assets
```

Reply:
@robbkidd I'm using an anonymous volume in order to be able to run `yarn install` in the build stage of the Dockerfile. So locally the `/node_modules` folder does not sync up to the container `node_modules`. Won't your suggestion still have the issue of the assets getting compiled nicely in a separate container, but the web container won't be able to see them at all?

Reply:
🤔 Probably.

One option could be to unanonymize (nymize?) a volume for `node_modules` and share it between the web and asset services.

```
  app: &app
    build: .
    image: rubyforgood/circulate
    volumes:
      - .:/usr/src/app
      - node_modules:/usr/src/app/node_modules
    env_file:
      - config/docker.env
    stdin_open: true
    tty: true
    depends_on:
      - database

  web:
    <<: *app
    ports:
      - '127.0.0.1:3000:3000'
      - '127.0.0.1:4000:4000'
    depends_on:
      - database
      - assets

  assets:
    <<: *app
    command: bin/webpack-dev-server

volumes:
  db_data:
  node_modules:
```

(Er, that has a bunch of my tinkering included.)

Like the anonymous volume, if the named volume is empty, the `node_modules` from the image build will be used. Once something changes in the containers so that `node_modules` is written to, both containers will use that new content.

Reply:
> If you use exec, you don't need to stop the terminal to run the DB migrations or db:reset. 

Unfortunately this is not true:

```
➭ docker-compose exec web rails db:reset
I, [2020-09-23T12:53:34.731258 #278]  INFO -- Mjml: Using mjml binary at /usr/src/app/node_modules/.bin/mjml
PG::ObjectInUse: ERROR:  database "circulate_development" is being accessed by other users
DETAIL:  There is 1 other session using the database.
Couldn't drop database 'circulate_development'
rails aborted!
ActiveRecord::StatementInvalid: PG::ObjectInUse: ERROR:  database "circulate_development" is being accessed by other users
DETAIL:  There is 1 other session using the database.
```

```
database_1         | 2020-09-23 12:53:41.100 UTC [66] ERROR:  database "circulate_development" is being accessed by other users
database_1         | 2020-09-23 12:53:41.100 UTC [66] DETAIL:  There is 1 other session using the database.
database_1         | 2020-09-23 12:53:41.100 UTC [66] STATEMENT:  DROP DATABASE IF EXISTS "circulate_development"
```

Reply:
Ah right, sorry. You can keep it up when you run db migrations, but not `db:reset`. Are you running `db:reset` every time you pull code in case there's been a change to the seeds file...? Why not just run the migrations (or `bin/update`)?

Reply:
@robbkidd @benlangfeld I tried to use the named volume approach suggested by @robbkidd:

```
services:
  web:
    build: .
    image: circulate/web
    volumes:
      - .:/usr/src/app
      - node_modules:/usr/src/app/node_modules
    env_file:
      - config/docker.env
    ports:
      - '3000:3000'
      - '4000:4000'
    depends_on:
      - database
      - assets
    stdin_open: true
    tty: true

  assets:
    image: circulate/web
    volumes:
      - .:/usr/src/app
      - node_modules:/usr/src/app/node_modules
    env_file:
      - config/docker.env
    command: bin/webpack-dev-server

...

volumes:
  db_data:
  node_modules:
```

which initially looks like it works, but in the logs, I see that the web app is still compiling the assets into the local `/node_modules` folder (which is not the same data as `/usr/src/app/node_modules`). I'm confused why that is happening. Any ideas?

Reply:
Actually, I don't think using a shared named volume will work unless we compile the assets after the container is created since currently they are getting compiled in the build stage of the Dockerfile. Using the anonymous volume gets around the local `node_modules` overriding the container `node_modules`, but I don't think using the shared name volume will work with the current Dockerfile configuration.